### PR TITLE
Perform Variable CSE on Variables of small enough size

### PR DESCRIPTION
**Pull request by jfix71**
This is an alternative to https://github.com/pytorch/glow/pull/1235

For Variables that are 256 elements or smaller and of the same type, are never written into, are private, and have no train kind, this will attempt to combine them into a single shared variable. This is particularly useful for IntLookupTables of the same "kind" (e.g. Tanh lookup tables), which often use the same data as their lookup table Var. Note that 256 is a heuristic used to ensure compile time isn't too long.

Also, note that I needed to add this after all other passes are complete, and not part of `optimize()` -- I added it to `optimize()` initially, but there are potential issues with other passes adding in new nodes that write to Variables. For example, `lower()` was creating new Save nodes for Variables that had been combined when lowering BatchNormalizationGrad, which breaks the program.

I haven't added a test yet -- if this is preferred to https://github.com/pytorch/glow/pull/1235 then I will add one.

### CODE REVIEW

**Comment by jfix71** [comment]
I stripped out the Visitor pattern as I'm not sure it's really useful in this case, can go back to it if preferred though.

Also considering renaming this to `mergeVars()` or something, as it's not working on subexpressions so CSE doesn't really make sense as a name.

**Comment by nadavrot** [comment]
Jordan, thanks for working on this. Every optimization is special in some way but we try to fit all of them into the standard pipe. What is exactly the problem with adding CSE to the opt pipeline? Optimizations need to preserve the semantics of the program so I am not sure why CSEing breaks the code. Are we only applying CSE to const vars? 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by jfix71** [comment]
@nadavrot If I add it to the end of `optimize()`, there is a problem when optimizing the lowered version of some gradient nodes, e.g. `BatchNormalizationGrad` and `SGD`. However I'm realizing this is probably due to a bug in NodeGen of not marking overwritten input correctly.

The lowering for `BatchNormalizationGrad` includes adding two `SaveNodes` to the graph which write to variables which previously were not written into; they were just inputs into `BatchNormalizationGrad` (mean and variance). `BatchNormalizationGradNode` does not mark these inputs as an overwritten. SGD also does not mark `Gsum` as overwritten even though it may be written to by a Save node too after lowering.

I suppose that if these two things are fixed, and we are sure we are not going to add other Nodes that write into Vars into the graph, it should be OK to add it back into `optimize()`.

**Comment by bertmaher** [comment]
With respect to compile time, would it make sense to hash the variable contents as well as the type?

**Comment by jfix71** [comment]
@bertmaher Good suggestion, I think it makes sense to hash some constant number of values, as it's likely the contents diverge very quickly if they're not equal. This way if they're not equal you don't have to iterate over all the contents to calculate the hash. Experimentally this was maybe 1% faster than either hashing all contents or none at all. Doesn't have a large impact because of the heuristic that only applies this to variables smaller than 256 elements. But if we decide it makes sense to increase 256 in the future then I would imagine hashing all the contents would grow comparatively worse.

**Comment by jfix71** [comment]
I've updated this by adding in the missing overwritten inputs to NodeGen, a test case, and the first 8 elements of the Variable's contents to the hash.

File: lib/Optimizer/GraphOptimizer.cpp [hunk 1]
```diff
@@ -1199,14 +1199,112 @@ struct CSEVisitor : NodeWalker {
...
+  for (auto &V : M->getVars()) {
+    // Perform CSE on all vars that have no writers.
+    if (varsWithWriters.find(V) != varsWithWriters.end()) {
+      continue;
+    }
+
+    // Only look for var CSE on vars of small enough size. Otherwise just skip
+    // them. 256 is used as a heuristic, to keep compile time reasonable.
+    size_t numEls = V->getType()->size();
+    if (numEls > 256) {
```
Comment (opti-mix) [inline lib/Optimizer/GraphOptimizer.cpp]: Could you introduce a named constexpr for 256 instead of hard-coding it here.

File: lib/Optimizer/GraphOptimizer.cpp [hunk 2]
```diff
@@ -1199,14 +1199,112 @@ struct CSEVisitor : NodeWalker {
...
 
+/// Common Subexpression Elimination for Variables.
+static void varsCSE(Module *M) {
+  // Create a set of Vars that have writers.
+  std::unordered_set<Node *> varsWithWriters;
+
+  // Determine which vars are written into, either as a result, or as an
+  // overwritten input. This needs to be done across all functions, as this
+  // combines variables which may be used across multiple functions.
+  for (auto *F : M->getFunctions()) {
```
Comment (opti-mix) [inline lib/Optimizer/GraphOptimizer.cpp]: Move this for loop and nested ones into a separate function?

File: lib/Optimizer/GraphOptimizer.cpp [hunk 3]
```diff
@@ -1199,14 +1199,112 @@ struct CSEVisitor : NodeWalker {
...
 };
 
+struct VarsHasherCSE {
+  size_t operator()(Variable *V) const {
+    auto hash = llvm::hash_value(V->getType());
+    auto &T = V->getPayload();
+    // Only use the first 8 elements in the hash. It's likely that if two
+    // tensors have different content they will diverge quickly. Fall back to
+    // full equality check in VarsEqCSE.
+    size_t numEls = std::min(T.getType().size(), 8ul);
```
Comment (opti-mix) [inline lib/Optimizer/GraphOptimizer.cpp]: Could you introduce a named constexpr for 8 instead of hard-coding it here.

**Comment by jfix71** [comment]
@qcolombet There are a couple problems with that approach currently. One, this optimization only applies to constant Variables, but Variables don't have a Mutability kind. So we don't know if a Variable is const until the entire graph is constructed, when we can then inspect to see if anyone writes into it.

Two, `createVariable()` doesn't take the Variable's data -- the way it's currently done is basically:
```
auto *var = createVariable(ElemKind::FloatTy, {3}, "name");
var->getHandle<float>() = {1,2,3};
```

Maybe we should consider adding a Mutability kind to Variable, and a `createVariable()` that takes the data of the Variable. I'm not sure that makes this PR obsolete, though.

**Comment by jfix71** [comment]
I squashed the commits and made requested changes. Happy to address any more issues/continue the discussion if this is the best path forward/what is otherwise.

**Comment by qcolombet** [comment]
> @qcolombet There are a couple problems with that approach currently. One, this optimization only applies to constant Variables, but Variables don't have a Mutability kind. So we don't know if a Variable is const until the entire graph is constructed, when we can then inspect to see if anyone writes into it.

I agree that this only makes sense for variables with data, not necessarily constants.
If we modify the data at compile time, we can create a new variable if it is used somewhere else.
If we modify the data at runtime, then we have some save nodes and that just means that we need to schedule these save nodes after the (soon to land™) load nodes.

> Two, createVariable() doesn't take the Variable's data -- the way it's currently done is basically:
>
> auto *var = createVariable(ElemKind::FloatTy, {3}, "name");
> var->getHandle<float>() = {1,2,3};
> Maybe we should consider adding a Mutability kind to Variable, and a createVariable() that takes the data of the Variable.

I am actually thinking that we could, like the loader do, have an API that creates a variable from a tensor. E.g., for this example `auto *var = createVariable("name", tensor123);`. At this point, this is easy to check if the variable needs to be created or not. Most of our code does it the other way around like you showed, so that means a lot of refactoring. Even if we would stick to the createVariable then getHandle to set the value, I am confident we could find a way to still share the variables at build time.

>  I'm not sure that makes this PR obsolete, though.

I have to admit I haven't looked at the PR, so I cannot really comment on that front :P. That said, if we only merge identical constant Variable, I believe we shouldn't need an optimization.

Anyhow, my main concern is that we are going to pay some compile time to solve an issue we first create and the worse part is that we would only do it for "small" tensor otherwise that's too expensive. That's why I am trying to see if there are not more efficient alternative.

The bottom line, I don't want to hold on this if that already giving some benefits!

File: lib/Optimizer/GraphOptimizer.cpp [hunk 4]
```diff
@@ -1199,14 +1199,117 @@ struct CSEVisitor : NodeWalker {
...
+      continue;
+    }
+
+    // Try to find a var that has the same data as the current one.
+    auto foundI = cseVars.find(V);
+    if (foundI == cseVars.end()) {
+      // No node CSE-equivalent to the current one has been seen yet. Remember
+      // this variable, so that the next occurrence can be replaced by this
+      // one.
+      cseVars.insert({V, V});
```
Comment (rdzhabarov) [inline lib/Optimizer/GraphOptimizer.cpp]: super nit: emplace would avoid unnecessary copies.

File: lib/Optimizer/GraphOptimizer.cpp [hunk 5]
```diff
@@ -1199,14 +1199,117 @@ struct CSEVisitor : NodeWalker {
...
+      assert(cseVars.find(V) != cseVars.end());
```
Comment (rdzhabarov) [inline lib/Optimizer/GraphOptimizer.cpp]: why is this needed? (we just inserted V into the map)

Comment (opti-mix) [reply lib/Optimizer/GraphOptimizer.cpp]: @rdzhabarov I think it is mostly copy/pasted from the CSE pass that I've written a while ago. IIRC, the initial idea of doing such a defensive check was to make sure that our custom hash and equality functions work properly when the insertion is performed. There were some issues with that at the beginning.

File: tools/ClassGen/NodeBuilder.h [hunk 6]
```diff
@@ -142,6 +145,18 @@ class NodeBuilder {
                      "input of a node");
   }
 
+  NodeBuilder &addOverwrittenGradientInput(const std::string &name) {
```
Comment (nadavrot) [inline tools/ClassGen/NodeBuilder.h]: Doxygen comment please.

**Comment by jfix71** [comment]
> I agree that this only makes sense for variables with data, not necessarily constants.
> If we modify the data at compile time, we can create a new variable if it is used somewhere else.
> If we modify the data at runtime, then we have some save nodes and that just means that we need to schedule these save nodes after the (soon to land™) load nodes.

Yeah, I meant constant at runtime, but I agree that once the load nodes land we should be able to support this if they have a single writer.

> I am actually thinking that we could, like the loader do, have an API that creates a variable from a tensor. E.g., for this example `auto *var = createVariable("name", tensor123);`.

Yeah I'm a fan of that. If we don't do this we would need some way to signal that the Variable is complete (e.g. `var->finalize()` or something) to know when it's OK to try to combine it with others. 

However, I don't think doing this at build time would make any difference for overall compile time compared to the optimization from this PR (assuming build time is considered part of compile time). This PR checks if the data is the same in tensors of 256 elements or less. We would need to perform essentially this same logic at build time, checking if the data in other Variable's tensors is the same.

The main way I think we can avoid the compile time overhead is doing something like what I had done in the alternative I put up earlier this week with singleton variables https://github.com/pytorch/glow/pull/1235. This only works if we know ahead of time the Variable we're repeatedly creating has the same data. Otherwise I don't see how we can avoid having to compare the data in each Variable's tensor.

File: lib/Optimizer/GraphOptimizer.cpp (hunk 5 above)
Comment (jfix71) [reply lib/Optimizer/GraphOptimizer.cpp]: Yep, what @opti-mix said. I can remove it if desired, but I did think it was useful.

File: tools/ClassGen/NodeGen.cpp [hunk 7]
```diff
@@ -393,6 +395,7 @@ int main(int argc, char **argv) {
       .addMember(MemberType::Float, "Momentum")
       .addMember(MemberType::Unsigned, "BatchSize")
       .addOverwrittenInput("Weight")
+      .addOverwrittenInput("Gsum")
```
Comment (qcolombet) [inline tools/ClassGen/NodeGen.cpp]: This should be done as a separate PR IMHO.

File: tools/ClassGen/NodeGen.cpp [hunk 8]
```diff
@@ -125,6 +125,8 @@ int main(int argc, char **argv) {
       .addMember(MemberType::Float, "Epsilon")
       .addMember(MemberType::Float, "Momentum")
       .addResult("Input.getType()")
+      .addOverwrittenGradientInput("Mean")
+      .addOverwrittenGradientInput("Var")
```
Comment (qcolombet) [inline tools/ClassGen/NodeGen.cpp]: This is only true for when we compile for training, thus I am not sure we want to put that unconditionally.
Put differently, although that's correct, I wonder if we are not leaving performance on the table for inference, which is what we care the most.

What do you think?

Comment (jfix71) [reply tools/ClassGen/NodeGen.cpp]: This is `addOverwrittenGradientInput()`, not `addOverwrittenInput()`, so it does not impact the non-gradient node `BatchNormalizationNode`. It only marks `Mean` and `Var` as overwritten  in `BatchNormalizationGradientNode`. So it should not change anything/negatively impact performance for inference.

Comment (qcolombet) [reply tools/ClassGen/NodeGen.cpp]: Ah right, but that's not accurate, isn't it?

I mean those inputs are going to be modified by the lowering of BN when we are in training mode not just by the lowering of BNG. (Admittedly, having one without the other would be pretty strange.)

File: lib/Optimizer/GraphOptimizer.cpp [hunk 9]
```diff
@@ -1219,14 +1219,119 @@ struct CSEVisitor : NodeWalker {
...
+    for (unsigned i = 0; i < N->getNumInputs(); i++)
+      if (N->isOverwrittenNthInput(i))
+        if (V == N->getNthInput(i))
+          return true;
+  }
+  return false;
+}
+
+/// Common Subexpression Elimination for Variables in the Module \p M.
+static void varsCSE(Module *M) {
```
Comment (nadavrot) [inline lib/Optimizer/GraphOptimizer.cpp]: I think that a better name would be "deduplicate constants".

Comment (jfix71) [reply lib/Optimizer/GraphOptimizer.cpp]: Done.

**Comment by jfix71** [comment]
FYI we can now revisit this since https://github.com/pytorch/glow/pull/1326 fixes the BN issue and has landed.

File: tests/unittests/graphOptzTest.cpp [hunk 10]
```diff
@@ -1114,3 +1114,39 @@ TEST_F(GraphOptz, concatReshapes) {
...
+  auto *CN = F_->createConcat("concat", {TN, SN, RN}, /* axis */ 0);
+  F_->createSave("ret", CN);
+
+  // Initially there are inputs 1, 2, and 3, and the variable for the save.
+  EXPECT_EQ(mod_.getVars().size(), 4);
+
+  ::glow::optimize(F_, CompilationMode::Infer);
+
+  // Now input1 and input2 have been combined, but input 3 has not.
+  EXPECT_EQ(mod_.getVars().size(), 3);
```
Comment (qcolombet) [inline tests/unittests/graphOptzTest.cpp]: Could you check that we are indeed left with the variables we expect?
Also, check that the users have been properly updated.

Comment (jfix71) [reply tests/unittests/graphOptzTest.cpp]: Done.

File: lib/Optimizer/GraphOptimizer.cpp [hunk 11]
```diff
@@ -1219,14 +1219,121 @@ struct CSEVisitor : NodeWalker {
...
+static void deduplicateConstants(Module *M) {
+  // Map from Variables to other Variables that are equivalent for purposes of
+  // deduplication.
+  std::unordered_map<Variable *, Variable *, VarsHasherDedup, VarsEqDedup>
+      duplicateVars;
+
+  for (auto &V : M->getVars()) {
+    // Only perform deduplication on vars of small enough size. Otherwise just
+    // skip them. 256 is used as a heuristic, to keep compile time reasonable.
+    constexpr size_t maxNumEls = 256;
```
Comment (qcolombet) [inline lib/Optimizer/GraphOptimizer.cpp]: Could we make this a cl::opt with the default being 256?
That way we can easily play with different sizes for tuning or if we have a bug we can easily rule that optimization out by putting maxNumEls to 0.

Comment (jfix71) [reply lib/Optimizer/GraphOptimizer.cpp]: Done.

**Comment by nadavrot** [comment]
Nice!
